::: {layout-ncol="2"}
[![](https://www.wellcomeconnectingscience.org/wp-content/themes/new_wellcome_connecting_science/dist/assets/svg/logo.svg){fig-align="left" width="300"}](https://coursesandconferences.wellcomeconnectingscience.org/event/single-cell-genomic-approaches-to-study-the-immune-system-latin-america-the-caribbean-20241109/)

[![](https://github.com/EveliaCoss/Tutorial_ISCB_LATAM_scATACseq/blob/main/Logo_JAGUAR.png?raw=true){fig-align="right" width="150"}](https://jaguar.liigh.unam.mx/)
:::

# Global information {.unnumbered}

::: {.panel-tabset group="globalInfo"}
### About the course üìå

This course is part of the [Workshop Single Cell Genomic Approaches to Study the Immune System - Latin America & the Caribbean](https://coursesandconferences.wellcomeconnectingscience.org/event/single-cell-genomic-approaches-to-study-the-immune-system-latin-america-the-caribbean-20241109/) (9-11 Novembrer 2024), Universidad CES, Medell√≠n, Colombia.

\- **Date:** 11 November 2024

\- **Course duration:** 5 hours

#### **Speakers:**

-   **Diego Ramirez** - Bachelor in Genomic Sciences, Escuela Nacional de Estudios Superiores Unidad Juriquilla UNAM (ENES Juriquilla), Mexico.

-   **Evelia Lorena Coss-Navarrete** - PostDoc, International Laboratory for Human Genome Research (LIIGH)-UNAM, Mexico. Contact: [Web page](https://eveliacoss.github.io/)

#### **Abstract:**

Single-cell transposase-accessible chromatin sequencing (scATAC-seq) represents the most innovative technology for examining genome-wide regulatory landscapes in single cells. For this tutorial, we will be analyzing a single-cell ATAC-seq dataset of human peripheral blood mononuclear cells (PBMCs) provided by 10x Genomics. We will run bridge integration for PBMC with the newly released Azimuth ATAC workflow. In this workshop we will review the existing statistical tools for analyzing scATAC-seq data, how to document your analysis and review some tools for interpreting results.

#### **Learning objectives:**

-   1\. Fundamentals of Single Cell ATAC-seq (scATAC-seq) and Multiome analysis.

-   2\. Single-Cell ATAC-seq Pre-Processing and Quality Control.

-   3\. Loading in and pre-processing the scATAC-seq, multiome, and scRNA-seq reference datasets.

-   4\. Mapping the scATAC-seq dataset via bridge integration.

-   5\. Exploring and assessing the resulting annotations.

-   6\. Motif analysis with Signac

### Agenda üìÜ {#agenda}

+-------------------------------------------------------------------+----------------------+-------------------------------+
| Subject                                                           | Time                 | Instructor                    |
+===================================================================+:====================:+:=============================:+
| üî∑ **Section 1 (45 min) - (9:45-10:30 h)**                        |                      |                               |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| Fundamentals of Single Cell ATAC-seq (scATAC-seq) analysis        | 15 min (9:45- 10:00) | Evelia Coss                   |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| Settings üíª                                                       | 10 min\              | Evelia Coss                   |
|                                                                   | (10:00-10:10)        |                               |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| Pre-Processing and Quality Control, gene activity matrix approach | 20 min\              | Evelia Coss                   |
|                                                                   | (10:10-10:30)        |                               |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| ‚òï *Coffee break (30 min) - (10:30-11:00 h) ‚òï*                   |                      |                               |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| üî∑¬†**Section 2 (2 h) - (11:00-13:00 h)**                          |                      |                               |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| Downtream analysis and multiomic dataset                          | 1 h\                 | Diego Ramirez                 |
|                                                                   | (11:00-12:00)        |                               |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| Mapping the scATAC-seq dataset via bridge integration             | 1 h\                 | Diego Ramirez                 |
|                                                                   | (12:00-13:00)        |                               |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| *üçñüçó Lunch (1 h) - (13:00 - 14:00) üçñüçó*                         |                      |                               |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| üî∑¬†**Section 3 (14:00 - 15:30 h)**                                |                      |                               |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| Motif analysis with Signac                                        | 30 min (14:00-14:30) | Diego Ramirez                 |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| Exploring results (graphs)                                        | 30 min (14:30-15:00) | Evelia Coss                   |
+-------------------------------------------------------------------+----------------------+-------------------------------+
| Other tools used in scATAC-seq                                    | 30 min (14:30-15:00) | Diego Ramirez and Evelia Coss |
+-------------------------------------------------------------------+----------------------+-------------------------------+

### About datasets **üìñ**

In this tutorial, we go over how to use scvi-tools functionality in R for analyzing ATAC-seq data. We will closely follow the PBMC tutorial from \[Signac\](https://stuartlab.org/signac/articles/pbmc_vignette#computing-qc-metrics), using \[scvi-tools\](https://docs.scvi-tools.org/en/1.0.4/tutorials/notebooks/atac/peakvi_in_R.html) when appropriate. In particular, we will

-   1\. Use PeakVI for dimensionality reduction and differential accessiblity for the ATAC-seq data

-   2\. Use scVI to integrate the unpaired ATAC-seq dataset with a match scRNA-seq dataset of PBMCs

This tutorial requires Reticulate. Please check out our installation guide for instructions on installing Reticulate and scvi-tools.

### About packages üìö

\[rmdformats\](https://github.com/juba/rmdformats)

Depending on the features provided by the template, you can add the following options to your YAML preamble. Look at the template function help page for a valid list :

fig_width : figures width, in inches

fig_height : figures height, in inches

fig_caption : toggle figure caption rendering

highlight : syntax highlighting

thumbnails : if TRUE, display content images as thumbnails

lightbox : if TRUE, add lightbox effect to content images

gallery : if TRUE, add navigation between images when displayed in lightbox

use_bookdown : if TRUE, will use bookdown instead of rmarkdown for HTML rendering, thus providing section numbering and cross references.

embed_fonts : if TRUE (default), use local files for fonts used in the template instead of links to Google Web fonts. This leads to bigger files but ensures that the fonts are available

additional aguments are passed to the base html_document RMarkdown template
:::
