# Exercises {#install .setup style="blue"}

## Dataset overview: **1k Peripheral Blood Mononuclear Cells (PBMCs) from a Healthy Donor**

For this tutorial, we will be analyzing a single-cell ATAC-seq dataset of [human peripheral blood mononuclear cells (PBMCs)](https://www.10xgenomics.com/datasets/1-k-peripheral-blood-mononuclear-cells-pbm-cs-from-a-healthy-donor-next-gem-v-1-1-1-1-standard-1-1-0) provided by 10x Genomics (Next GEM v1.1). The following files are used in this course, all available through the 10x Genomics website:

-   The [Raw data](https://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_pbmc_1k_nextgem/atac_pbmc_1k_nextgem_filtered_peak_bc_matrix.h5): Filtered peak barcode matrix.

-   The [Metadata](https://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_pbmc_1k_nextgem/atac_pbmc_1k_nextgem_singlecell.csv): Per-barcode fragment counts & metrics.

-   The [fragments file](https://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_pbmc_1k_nextgem/atac_pbmc_1k_nextgem_fragments.tsv.gz): Barcoded and aligned fragment file.

-   The [fragments file index](https://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_pbmc_1k_nextgem/atac_pbmc_1k_nextgem_fragments.tsv.gz.tbi): Fragments file index.

::: {.callout-note icon="false"}
## Questions

Check the [data report](https://cf.10xgenomics.com/samples/cell-atac/1.1.0/atac_pbmc_1k_nextgem/atac_pbmc_1k_nextgem_web_summary.html) and answer the questions.

-   How many cells were recovered?

    ::: {.callout-tip collapse="true" icon="false"}
    #### Answer

    -   1,195 nuclei were recovered.
    :::

-   Which version of Cell Ranger ATAC is used?

    ::: {.callout-tip collapse="true" icon="false"}
    #### Answer

    -   Single Cell ATAC dataset analyzed using Cell Ranger ATAC 1.1.0
    :::

-   What version of the genome did they use?

    ::: {.callout-tip collapse="true" icon="false"}
    ## Answer

    -   hg19
    :::
:::

#### **Pre-processing workflow**

```{mermaid}
%%| fig-width: 15 
flowchart LR    
  A(Import Data) --> B(Annotation)     
  B --> C(Computing QC metrics)     
  C --> D(Normalization and linear \ndimensional reduction)      
  D --> E(Non-linear dimensional \nreduction and clustering) 
```

::: callout-note
## Exercise in Google Colab

1.  Create a copy of the [Google colab notebook part 13](https://colab.research.google.com/drive/1o-2mHth0v4zvTxSxI6y_RQBKNiw50QlK?usp=sharing) on your one drive.
2.  Run the exercise.
:::

## References

-   Signac tutorial - [Analyzing PBMC scATAC-seq (Signac tutorial)](https://stuartlab.org/signac/articles/pbmc_vignette)
-   Signac tutorial - [Analyzing PBMC scATAC-seq](https://rpubs.com/qshenfeng/signac_pbmc10k)
