# **Practical 17:** Exploring results (graphs) {.setup style="blue"}

## The number of cells in each cell-type biases the null distributions and statistics of the Z-scores method

This study compares Z-scores, derived from the Pearson R value of cis-links between ATACseq peaks and nearby genes, to their matched trans-link null distribution. After analyzing PBMC multiomic data, significant differences were found between the Pearson R coefficients and Z-scores for many peak-gene links. An example is an ATACseq peak near the *NOD2* gene, correlated with monocytes (R = 0.12), but showing no significant link using Z-scores (P = 0.07). However, a significant link was found with *SNX20* (P = 0.02). Bimodal null distributions for trans-peaks at the *NOD2* locus make Z-scores inaccurate. Excluding specific cell-type peaks leads to a unimodal distribution, highlighting the impact of peak selection on Z-scores.

[![The Z-scores method misses candidate regulatory sequences linked to NOD2 expression in peripheral blood mononuclear cells (PBMC). (A) The Z-scores model matches an ATACseq peak for GC content and coverage with ATACseq peaks in trans to create a scaled null distribution, producing Z-scores for each trans-links and the tested peak. (B) ATACseq tracks at the NOD2 locus identified in PBMC. The grey areas (labeled 1--2-3) highlight the top three ATACseq peaks correlations with NOD2 expression using the simple Pearson R method. Peak #1 (chr16-50,684,843--50,685,984) includes an eQTL for NOD2 that is also associated with leprosy and Crohn's disease by GWAS. The loops highlighted in the "Links" row are identified using the Z-score method (P-value \< 0.05); we note that there is no significant link between peak chr16-50,684,843--50,685,984 (peak #1) and NOD2. Loops are drawn from the middle of the ATACseq peaks to the transcription start site of the correlated gene(s). In the right column, we showed (top to bottom) RNAseq UMAP of cell-type annotations, SNX20 expression density, NOD2 expression density, and chr16-50,684,843--50,685,984 ATACseq accessibility density. The violin plots represent NOD2 expression levels in each cell-type. Leblanc & Lettre, et al. 2023. Scientific reports.](img/Figure1_example.png){#fig-example1}](https://www.nature.com/articles/s41598-023-31040-w/figures/1)

## Single cell lineage determination

[![Lineage determination of single cells from skin, kidney, and peripheral blood mononuclear cells (PBMCs). (A) Clustering of cells (n = 899) by t-distributed stochastic neighbor embedding (t-SNE). Cells are colored based on tissue of origin from skin (blue), kidney cells (red), and PBMCs (yellow). (B) Six distinct clusters generated by t-SNE plotting. (C) Differentially expressed genes across 6 cell clusters. In this heat map, rows correspond to individual genes found to be selectively upregulated in individual clusters (P \< 0.01). (D) Violin plots demonstrating expression of lineage markers that indicate the identity of the clusters generated by t-SNE plotting. Der, et al. 2017. JCI insight.](https://df6sxcketz7bb.cloudfront.net/manuscripts/93000/93009/medium/jci.insight.93009.f3.jpg){#fig-lupus-example}](https://insight.jci.org/articles/view/93009)

## **Single-cell chromatin accessibility reveals principles of regulatory variation**

[![Structured cis-variability across single epigenomes. a, Per-cell deviations of expected fragments across a region within chromosome 1, For display, only large deviation cells are shown (n = 186 cells). b, Pearson correlation coefficient representing chromosome compartment signal of interaction frequency from a chromatin conformation capture assay (left, analysis carried out of data from ref. 27) or doubly correlated normalized deviations of scATAC-seq (right) from chromosome 1. Data in white represents masked regions due to highly repetitive regions. c, Permuted cis-correlation map for chromosome 1 (analysed identically to b). d, Box highlights a representative region depicting long-range covariability. Buenrostro, et al. 2015. Nature](https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fnature14590/MediaObjects/41586_2015_Article_BFnature14590_Fig4_HTML.jpg?as=webp){#fig-epigenomic}](https://www.nature.com/articles/nature14590)

# CITE-seq and scATAC-seq

To learn more about how the antibody barcode matrix is computationally generated from the sequencing data, please visit [CITE-seq-Count](https://hoohm.github.io/CITE-seq-Count/). To learn more about CITE-seq and feature barcoding, please visit the [CITE-seq site.](https://cite-seq.com/)

## **References**

-   Signac - tutorial [Building trajectories with Monocle 3](https://stuartlab.org/signac/articles/monocle)
-   Signac - tutorial [Visualization of genomic regions](https://stuartlab.org/signac/1.13.0/articles/visualization)
-   Leblanc & Lettre, et al. 2023. [Major cell-types in multiomic single-nucleus datasets impact statistical modeling of links between regulatory sequences and target genes](https://www.nature.com/articles/s41598-023-31040-w). Scientific reports
-   Buenrostro, et al. 2015. [Single-cell chromatin accessibility reveals principles of regulatory variation](https://www.nature.com/articles/nature14590). Nature.
-   Orr Ashenber, et al. Tutorial. [Analysis of single cell RNA-seq data](https://broadinstitute.github.io/2019_scWorkshop/)
-   Der, et al. 2017.[Single cell RNA sequencing to dissect the molecular heterogeneity in lupus nephritis](https://insight.jci.org/articles/view/93009) JCI insight.
-   [CITE-seq and scATAC-seq tutorial.](https://broadinstitute.github.io/2019_scWorkshop/cite-seq-and-scatac-seq.html)
