# **Practical 17:** Exploring results (graphs) {#install .setup style="blue"}

[![](https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fnature14590/MediaObjects/41586_2015_Article_BFnature14590_Fig4_HTML.jpg?as=webp)](https://www.nature.com/articles/nature14590)

https://www.nature.com/articles/nature14590

# CITE-seq and scATAC-seq

https://broadinstitute.github.io/2019_scWorkshop/cite-seq-and-scatac-seq.html

# Building trajectories with Monocle 3

Also you can construct cell trajectories with Monocle 3 using single-cell ATAC-seq data. Please see the [Monocle 3 website](https://cole-trapnell-lab.github.io/monocle3/) for information about installing Monocle 3.

[![](https://stuartlab.org/signac/articles/monocle_files/figure-html/unnamed-chunk-18-2.png)](https://stuartlab.org/signac/articles/monocle)

## **References**

-   Signac - tutorial [Building trajectories with Monocle 3](https://stuartlab.org/signac/articles/monocle)
-   Signac - tutorial [Visualization of genomic regions](https://stuartlab.org/signac/1.13.0/articles/visualization)
