<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Analyzing PBMCs Single Cell ATAC- Seq (scATAC-Seq) and Multiome data: Theory and practice - 4&nbsp; Practical 14: scATAC-seq Downstream</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./Section2_P14_retos.html" rel="next">
<link href="./Section1_P13_retos.html" rel="prev">
<link href="./JAGUAR_logo.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="site_libs/quarto-diagram/mermaid.min.js"></script>
<script src="site_libs/quarto-diagram/mermaid-init.js"></script>
<link href="site_libs/quarto-diagram/mermaid.css" rel="stylesheet">


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./Section2_P14_intro.html"><strong>Section 2</strong> - scATAC-seq Downstream analysis</a></li><li class="breadcrumb-item"><a href="./Section2_P14_intro.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">**Practical 14:** scATAC-seq Downstream</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Analyzing PBMCs Single Cell ATAC- Seq (scATAC-Seq) and Multiome data: Theory and practice</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/EveliaCoss/Tutorial_ISCB_LATAM_scATACseq" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Share" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Share"><i class="bi bi-share"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://twitter.com/intent/tweet?url=|url|">
              <i class="bi bi-bi-twitter pe-1"></i>
            Twitter
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://www.facebook.com/sharer/sharer.php?u=|url|">
              <i class="bi bi-bi-facebook pe-1"></i>
            Facebook
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://www.linkedin.com/sharing/share-offsite/?url=|url|">
              <i class="bi bi-bi-linkedin pe-1"></i>
            LinkedIn
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Global information</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text"><strong>Section 1</strong> - Introduction and Quality control</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section1_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Fundamentals of Single Cell ATAC-seq (scATAC-seq) analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section1_P13_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title"><strong>Practical 13:</strong> scATAC-seq Pre-Processing and Quality Control</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section1_P13_retos.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text"><strong>Section 2</strong> - scATAC-seq Downstream analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section2_P14_intro.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title"><strong>Practical 14:</strong> scATAC-seq Downstream</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section2_P14_retos.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section2_P15_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title"><strong>Practical 15:</strong> Analyses and scRNA-seq Integration</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section2_P15_retos.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text"><strong>Section 3</strong> - Motif analysis and results</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section3_P16_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title"><strong>Practical 16:</strong> Motif analysis with Signac</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section3_P16_retos.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section3_P17_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title"><strong>Practical 17:</strong> Exploring results (graphs)</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section3_P17_retos.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Section3_P18_others.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Other tools used in scATAC-seq</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Summary</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#downstream-workflow" id="toc-downstream-workflow" class="nav-link active" data-scroll-target="#downstream-workflow"><strong>Downstream workflow</strong></a></li>
  <li><a href="#option-a-gene-activity-matrix-approach-rna-imputation" id="toc-option-a-gene-activity-matrix-approach-rna-imputation" class="nav-link" data-scroll-target="#option-a-gene-activity-matrix-approach-rna-imputation"><span class="header-section-number">4.1</span> Option A: Gene activity matrix approach / RNA imputation</a>
  <ul>
  <li><a href="#step-1-load-data" id="toc-step-1-load-data" class="nav-link" data-scroll-target="#step-1-load-data"><span class="header-section-number">4.1.1</span> Step 1: Load data</a></li>
  <li><a href="#step-2-create-a-expression-matrix" id="toc-step-2-create-a-expression-matrix" class="nav-link" data-scroll-target="#step-2-create-a-expression-matrix"><span class="header-section-number">4.1.2</span> Step 2: Create a expression matrix</a></li>
  <li><a href="#step-3-check-biomarkers" id="toc-step-3-check-biomarkers" class="nav-link" data-scroll-target="#step-3-check-biomarkers"><span class="header-section-number">4.1.3</span> Step 3: Check biomarkers</a></li>
  </ul></li>
  <li><a href="#signac-workflow" id="toc-signac-workflow" class="nav-link" data-scroll-target="#signac-workflow"><span class="header-section-number">4.2</span> Signac Workflow</a></li>
  <li><a href="#label-transfer" id="toc-label-transfer" class="nav-link" data-scroll-target="#label-transfer"><span class="header-section-number">4.3</span> <strong>Label transfer</strong></a>
  <ul>
  <li><a href="#step-4-integrating-with-scrna-seq-data-multimodal" id="toc-step-4-integrating-with-scrna-seq-data-multimodal" class="nav-link" data-scroll-target="#step-4-integrating-with-scrna-seq-data-multimodal"><span class="header-section-number">4.3.1</span> Step 4: <strong>Integrating with scRNA-seq data (multimodal)</strong></a></li>
  <li><a href="#step-5-find-transfer-anchors" id="toc-step-5-find-transfer-anchors" class="nav-link" data-scroll-target="#step-5-find-transfer-anchors"><span class="header-section-number">4.3.2</span> Step 5: Find transfer anchors</a></li>
  <li><a href="#step-7-annotate-scatac-seq-cells-via-label-transfer" id="toc-step-7-annotate-scatac-seq-cells-via-label-transfer" class="nav-link" data-scroll-target="#step-7-annotate-scatac-seq-cells-via-label-transfer"><span class="header-section-number">4.3.3</span> Step 7: Annotate scATAC-seq cells via label transfer</a></li>
  <li><a href="#step-8-remove-platelets" id="toc-step-8-remove-platelets" class="nav-link" data-scroll-target="#step-8-remove-platelets"><span class="header-section-number">4.3.4</span> Step 8: Remove platelets</a></li>
  <li><a href="#step-9-rename-labels" id="toc-step-9-rename-labels" class="nav-link" data-scroll-target="#step-9-rename-labels"><span class="header-section-number">4.3.5</span> Step 9: Rename labels</a></li>
  <li><a href="#step-10-compare-the-results" id="toc-step-10-compare-the-results" class="nav-link" data-scroll-target="#step-10-compare-the-results"><span class="header-section-number">4.3.6</span> Step 10: Compare the results</a></li>
  </ul></li>
  <li><a href="#find-differentially-accessible-peaks-between-cell-types" id="toc-find-differentially-accessible-peaks-between-cell-types" class="nav-link" data-scroll-target="#find-differentially-accessible-peaks-between-cell-types"><span class="header-section-number">4.4</span> Find differentially accessible peaks between cell types</a></li>
  <li><a href="#go-enrichment-analysis-with-clusterprofiler" id="toc-go-enrichment-analysis-with-clusterprofiler" class="nav-link" data-scroll-target="#go-enrichment-analysis-with-clusterprofiler"><span class="header-section-number">4.5</span> GO enrichment analysis with clusterProfiler</a></li>
  <li><a href="#plotting-genomic-regions" id="toc-plotting-genomic-regions" class="nav-link" data-scroll-target="#plotting-genomic-regions"><span class="header-section-number">4.6</span> Plotting genomic regions</a></li>
  <li><a href="#calling-peaks" id="toc-calling-peaks" class="nav-link" data-scroll-target="#calling-peaks"><span class="header-section-number">5</span> Calling peaks</a>
  <ul>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">5.1</span> References</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="install" class="quarto-section-identifier"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title"><strong>Practical 14:</strong> scATAC-seq Downstream</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>In this step, we will demonstrate the following:</p>
<ul>
<li>Utilizing predicted expression through the gene activity matrix approach.</li>
<li>Integrating ATAC-seq and RNA-seq data for label transfer using mutual nearest neighbor cell anchors.</li>
<li>Identifying differentially accessible peaks to assess chromatin accessibility variations across cell types.</li>
<li>Performing GO enrichment analysis with clusterProfiler to find overrepresented biological processes in gene sets.</li>
<li>Plotting genomic regions to visualize the distribution and features of specific sequences across the genome.</li>
</ul>
<section id="downstream-workflow" class="level4">
<h4 class="anchored" data-anchor-id="downstream-workflow"><strong>Downstream workflow</strong></h4>
<div class="cell" data-fig-width="15">
<div class="cell-output-display">
<div>
<div>
<pre class="mermaid mermaid-js">flowchart LR    
  A{Integrating with scRNA-seq data} --&gt; B(Gene activity matrix approach - \nRNA imputation)     
  A --&gt; C(Dictionary Learning for cross-modality \nintegration - Bridge integration)
  B --&gt; D[Check Biomarkers]
  C --&gt; D
</pre>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What if we don’t have the RNA information?
</div>
</div>
<div class="callout-body-container callout-body">
<p>The RNA information in the multiome data is great because it is much easier to annotate cell types/states by gene expression. However, what if we have scATAC-seq rather than scMultiome data? In that case is there a way to “predict” the gene expression based on the ATAC fragments? Although not perfect, there is indeed some possible solutions.</p>
</div>
</div>
</section>
<section id="option-a-gene-activity-matrix-approach-rna-imputation" class="level2 setup" style="blue" data-number="4.1">
<h2 class="setup anchored" style="blue" data-number="4.1" data-anchor-id="option-a-gene-activity-matrix-approach-rna-imputation"><span class="header-section-number">4.1</span> Option A: Gene activity matrix approach / RNA imputation</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://denovosoftware.com/faq/kb-how-can-i-explore-tsne-umap-plots/"><img src="https://denovosoftware.com/wp-content/uploads/Exploring-tSNE-parameter-overlays.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<p>Gene activity scores capture how much open chromatin there is in the promoter regions of each gene (by default <strong>2000 bp (2 kb) upstream</strong>). The assumption here is that open chromatin is a proxy for gene expression. Gene activity scores are represented as a matrix, with one row per gene and one column per cell. This makes the gene activitiy scores directly compatible with single cell RNA-seq data.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Calculating the gene activity scores takes around 10 minutes for 2000 cells and all genes.</p>
</div>
</div>
<p>We can try to quantify the activity of each gene in the genome by assessing the chromatin accessibility associated with the gene, and create a new gene activity assay derived from the scATAC-seq data. Here we will use a simple approach of summing the fragments intersecting the gene body and promoter region (we also recommend exploring the&nbsp;<a href="https://cole-trapnell-lab.github.io/cicero-release/">Cicero</a>&nbsp;tool, which can accomplish a similar goal, and we provide a vignette showing how to run Cicero within a Signac workflow&nbsp;<a href="https://stuartlab.org/signac/articles/cicero">here</a>).</p>
<section id="step-1-load-data" class="level3" data-number="4.1.1">
<h3 data-number="4.1.1" class="anchored" data-anchor-id="step-1-load-data"><span class="header-section-number">4.1.1</span> Step 1: Load data</h3>
<p>We then count the number of fragments for each cell that map to each of these regions, using the using the&nbsp;<a href="https://stuartlab.org/signac/reference/featurematrix"><code>FeatureMatrix()</code></a>&nbsp;function. These steps are automatically performed by the&nbsp;<a href="https://stuartlab.org/signac/reference/geneactivity"><code>GeneActivity()</code></a>&nbsp;function:</p>
</section>
<section id="step-2-create-a-expression-matrix" class="level3" data-number="4.1.2">
<h3 data-number="4.1.2" class="anchored" data-anchor-id="step-2-create-a-expression-matrix"><span class="header-section-number">4.1.2</span> Step 2: Create a expression matrix</h3>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="at">file =</span> <span class="st">"data/pbmc.RData"</span>) <span class="co"># scATAC</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>start <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>gene.activities <span class="ot">&lt;-</span> <span class="fu">GeneActivity</span>(pbmc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting gene coordinates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting reads overlapping genomic regions</code></pre>
</div>
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>end <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>end <span class="sc">-</span> start</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Time difference of 13.86905 mins</code></pre>
</div>
</div>
<p>Before:</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>pbmc<span class="sc">@</span>assays</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>$peaks
ChromatinAssay data with 165376 features for 9649 cells
Variable features: 165376 
Genome: 
Annotation present: TRUE 
Motifs present: FALSE 
Fragment files: 1 </code></pre>
</div>
</div>
<p>Add the gene activity matrix to the Seurat object as a new assay and normalize it.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>pbmc[[<span class="st">'RNA'</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(<span class="at">counts =</span> gene.activities)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># normalization</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">assay =</span> <span class="st">'RNA'</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">normalization.method =</span> <span class="st">'LogNormalize'</span>,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale.factor =</span> <span class="fu">median</span>(pbmc<span class="sc">$</span>nCount_RNA)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>After:</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>pbmc<span class="sc">@</span>assays</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>$peaks
ChromatinAssay data with 165376 features for 9649 cells
Variable features: 165376 
Genome: 
Annotation present: TRUE 
Motifs present: FALSE 
Fragment files: 1 

$RNA
Assay data with 19607 features for 9649 cells
First 10 features:
 PLCXD1, GTPBP6, PPP2R3B, SHOX, CRLF2, CSF2RA, IL3RA, SLC25A6, ASMTL,
P2RY8 </code></pre>
</div>
</div>
</section>
<section id="step-3-check-biomarkers" class="level3" data-number="4.1.3">
<h3 data-number="4.1.3" class="anchored" data-anchor-id="step-3-check-biomarkers"><span class="header-section-number">4.1.3</span> Step 3: Check biomarkers</h3>
<p>We are now able to visualize the activity of canonical biomarkers to guide our interpretation of scATAC-seq clusters. Although this new putative “scRNA-seq” experiment derived from scATAC-seq will be noisier than a canonical scRNA-seq experiment, it will still be useful. The noise arises from the assumption made when generating the gene activity matrix, which assumes a perfect correlation between promoter/ORF accessibility and gene expression—something that is not always the case.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(pbmc) <span class="ot">&lt;-</span> <span class="st">'RNA'</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> <span class="fu">c</span>(<span class="st">'MS4A1'</span>, <span class="st">'CD3D'</span>, <span class="st">'LEF1'</span>, <span class="st">'NKG7'</span>, <span class="st">'TREM1'</span>, <span class="st">'LYZ'</span>),</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pt.size =</span> <span class="fl">0.1</span>,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">max.cutoff =</span> <span class="st">'q95'</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncol =</span> <span class="dv">3</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="Section2_P14_intro_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li><p><strong>MS4A1 (CD20)</strong>: Marker for B lymphocytes.</p></li>
<li><p><strong>CD3D</strong>: Part of the T cell receptor complex, indicating T cell presence.</p></li>
<li><p><strong>LEF1</strong>: Transcription factor associated with T and B cell differentiation.</p></li>
<li><p><strong>NKG7</strong>: Expressed in NK cells and T lymphocytes, related to cytotoxicity.</p></li>
<li><p><strong>TREM1</strong>: Present in monocytes and microglia, associated with the inflammatory response.</p></li>
<li><p><strong>LYZ (lysozyme)</strong>: Expressed in macrophages, involved in defense against bacteria.</p></li>
</ol>
<p>These markers indicate the activation and function of various immune cells.</p>
</div>
</div>
</section>
</section>
<section id="signac-workflow" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="signac-workflow"><span class="header-section-number">4.2</span> Signac Workflow</h2>
<div id="fig-signac" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="img/FigureSignac.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;4.1: Single-cell chromatin analysis workflow with Signac.</figcaption>
</figure>
</div>
</section>
<section id="label-transfer" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="label-transfer"><span class="header-section-number">4.3</span> <strong>Label transfer</strong></h2>
<p>After calculating the gene activity scores, we can now integrate the ATAC-seq data with the RNA-seq data.</p>
<p>The process begins by identifying anchors, which are pairs of cells—one from ATAC-seq and one from RNA-seq. To achieve this, we project both datasets into a shared space and identify pairs of cells that are mutual nearest neighbors (MNNs), one from each dataset. These pairs are then filtered to retain the most reliable ones, which serve as the anchors.</p>
<p>These anchors allow us to project the ATAC-seq data onto the RNA-seq data, enabling the identification of cell type annotations for nearby cells. In this way, annotations from the RNA-seq data can be transferred to the ATAC-seq data, a method commonly known as <strong>label transfer</strong>.</p>
<section id="step-4-integrating-with-scrna-seq-data-multimodal" class="level3" data-number="4.3.1">
<h3 data-number="4.3.1" class="anchored" data-anchor-id="step-4-integrating-with-scrna-seq-data-multimodal"><span class="header-section-number">4.3.1</span> Step 4: <strong>Integrating with scRNA-seq data (multimodal)</strong></h3>
<p>To help interpret the scATAC-seq data, we can classify cells based on an <strong>scRNA-seq experiment from the same biological system</strong> (human PBMC). We utilize methods for cross-modality integration and label transfer, described <a href="https://doi.org/10.1016/j.cell.2019.05.031">here</a>, with a more in-depth tutorial <a href="https://satijalab.org/seurat/v3.0/atacseq_integration_vignette.html">here</a>. We aim to identify shared correlation patterns in the gene activity matrix and scRNA-seq dataset to identify matched biological states across the two modalities. This procedure returns a classification score for each cell for each scRNA-seq-defined cluster label.</p>
<div id="fig-anchros" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867419305598%3Fshowall%3Dtrue"><img src="https://www.cell.com/cms/10.1016/j.cell.2019.05.031/asset/a55ffc75-7bc8-4e9b-8106-3f328acd1d09/main.assets/gr1.jpg" class="img-fluid figure-img"></a></p>
<figcaption class="figure-caption">Figure&nbsp;4.2: Schematic Overview of Reference “Assembly” Integration in Seurat v3. From: Stuart, et al.&nbsp;2019. Cell.</figcaption>
</figure>
</div>
<p>Here we load a pre-processed scRNA-seq dataset for human PBMCs, also provided by 10x Genomics. You can download the raw data for this experiment from the 10x <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_10k_v3">website</a>, and view the code used to construct this object on <a href="https://github.com/satijalab/Integration2019/blob/master/preprocessing_scripts/pbmc_10k_v3.R">GitHub</a>. Alternatively, you can download the pre-processed Seurat object <a href="https://signac-objects.s3.amazonaws.com/pbmc_10k_v3.rds">here</a>.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the pre-processed scRNA-seq data for PBMCs</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>pbmc_rna <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/pbmc_10k_v3.rds"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>pbmc_rna <span class="ot">&lt;-</span> <span class="fu">UpdateSeuratObject</span>(pbmc_rna)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># free memory</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>            used   (Mb) gc trigger    (Mb)   max used    (Mb)
Ncells  12418522  663.3   33923381  1811.8   53005282  2830.8
Vcells 781627810 5963.4 1375216326 10492.1 1363593522 10403.4</code></pre>
</div>
</div>
</section>
<section id="step-5-find-transfer-anchors" class="level3" data-number="4.3.2">
<h3 data-number="4.3.2" class="anchored" data-anchor-id="step-5-find-transfer-anchors"><span class="header-section-number">4.3.2</span> Step 5: Find transfer anchors</h3>
<p>Find a set of anchors between a reference and query object. These anchors can later be used to transfer data from the reference to query object using the&nbsp;<a href="https://satijalab.org/seurat/reference/transferdata"><code>TransferData</code></a>&nbsp;object.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>transfer.anchors <span class="ot">&lt;-</span> <span class="fu">FindTransferAnchors</span>(</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">reference =</span> pbmc_rna, <span class="co"># scRNA</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">query =</span> pbmc, <span class="co"># scATAC</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">'cca'</span> <span class="co"># Perform dimensional reduction</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in size + sum(size_args, na.rm = FALSE): NAs produced by integer
overflow
Warning in size + sum(size_args, na.rm = FALSE): NAs produced by integer
overflow</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Running CCA</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Merging objects</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    Found 19133 anchors</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><br>
Dimensional reduction to perform when finding anchors. Options are:</p>
<ul>
<li><p><code>pcaproject</code>: Project the PCA from the reference onto the query. We recommend using Principal Component Analysis (PCA) when reference and query datasets are from scRNA-seq, because PCA effectively captures the variance in gene expression data.</p></li>
<li><p><code>lsiproject</code>: Project the LSI from the reference onto the query. We recommend using LSI when reference and query datasets are from scATAC-seq. This requires that LSI has been computed for the reference dataset, and the same features (eg, peaks or genome bins) are present in both the reference and query. See <strong><code>RunTFIDF</code></strong> and <strong><code>RunSVD</code>.</strong></p></li>
<li><p><code>rpca</code>: Project the PCA from the reference onto the query, and the PCA from the query onto the reference (reciprocal PCA projection). This bidirectional approach allows for a more comprehensive alignment of the datasets, potentially enhancing the identification of shared features.</p></li>
<li><p><code>cca</code>: Canonical Correlation Analysis (CCA) is used to find linear relationships between the reference and query datasets. By identifying the canonical components that maximize the correlation between the two datasets, CCA helps in aligning them based on shared information.</p></li>
</ul>
</div>
</div>
</section>
<section id="step-7-annotate-scatac-seq-cells-via-label-transfer" class="level3" data-number="4.3.3">
<h3 data-number="4.3.3" class="anchored" data-anchor-id="step-7-annotate-scatac-seq-cells-via-label-transfer"><span class="header-section-number">4.3.3</span> Step 7: Annotate scATAC-seq cells via label transfer</h3>
<p>After identifying anchors, we can transfer annotations from the scRNA-seq dataset into the scATAC-seq cells. The annotations are stored in the&nbsp;<code>seurat_annotations</code>&nbsp;field, and are provided as input to the&nbsp;<code>refdata</code>&nbsp;parameter. The output will contain a matrix with predictions and confidence scores for each ATAC-seq cell.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>predicted.labels <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">anchorset =</span> transfer.anchors,</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">refdata =</span> pbmc_rna<span class="sc">$</span>celltype,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">weight.reduction =</span> pbmc[[<span class="st">'lsi'</span>]], <span class="co"># reduction of the original `seurat` object's dim</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">30</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vector weights</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Predicting cell labels</code></pre>
</div>
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(<span class="at">object =</span> pbmc, <span class="at">metadata =</span> predicted.labels)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gc</span>()<span class="co"># free memory</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>            used   (Mb) gc trigger    (Mb)   max used    (Mb)
Ncells  12479518  666.5   33923381  1811.8   53005282  2830.8
Vcells 784162891 5982.7 1375216326 10492.1 1372394404 10470.6</code></pre>
</div>
</div>
<p><strong>Check plot</strong></p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc_rna,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">'celltype'</span>,</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'scRNA-seq'</span>)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>plot2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc,</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">'predicted.id'</span>,</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'scATAC-seq (prediction)'</span>)</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>plot1 <span class="sc">|</span> plot2 </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="Section2_P14_intro_files/figure-html/graph-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="step-8-remove-platelets" class="level3" data-number="4.3.4">
<h3 data-number="4.3.4" class="anchored" data-anchor-id="step-8-remove-platelets"><span class="header-section-number">4.3.4</span> Step 8: Remove platelets</h3>
<p>The scRNA-based classifications match the UMAP visualization from the scATAC-seq data. However, a small group of cells is unexpectedly predicted to be platelets, which lack nuclei and shouldn’t be detected by scATAC-seq. These cells might actually be megakaryocytes, platelet precursors usually found in the bone marrow but rarely in peripheral blood. Given the extreme rarity of megakaryocytes in normal bone marrow (&lt;0.1%), this seems unlikely.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(pbmc, <span class="st">'prediction.score.max'</span>, <span class="at">group.by =</span> <span class="st">'predicted.id'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="Section2_P14_intro_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Plotting the prediction score for the cells assigned to each label reveals that the “platelet” cells received relatively low scores (&lt; 0.8), indicating a low confidence in the assigned cell identity. In most cases, the next most likely cell identity predicted for these cells was “CD4 naive”.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the metadata columns that start with "prediction.score."</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>metadata_attributes <span class="ot">&lt;-</span> <span class="fu">colnames</span>(pbmc[[]])</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>prediction_score_attributes <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"^prediction.score."</span>, metadata_attributes, <span class="at">value =</span> <span class="cn">TRUE</span>)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>prediction_score_attributes <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(prediction_score_attributes, <span class="st">"prediction.score.max"</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the prediction score attributes for these cells</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>predicted_platelets <span class="ot">&lt;-</span> <span class="fu">which</span>(pbmc<span class="sc">$</span>predicted.id <span class="sc">==</span> <span class="st">"Platelet"</span>)</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>platelet_scores <span class="ot">&lt;-</span> pbmc[[]][predicted_platelets, prediction_score_attributes]</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Order the columns by their average values in descending order</span></span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>ordered_columns <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">sort</span>(<span class="fu">colMeans</span>(platelet_scores, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), <span class="at">decreasing =</span> <span class="cn">TRUE</span>))</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>ordered_platelet_scores_df <span class="ot">&lt;-</span> platelet_scores[, ordered_columns]</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ordered_platelet_scores_df)[<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                   prediction.score.CD4.Memory
ACAAAGAAGACACGGT-1                 0.025746481
CACTAAGGTAATGTAG-1                 0.008556831
CTCAACCAGCGAGCTA-1                 0.024011379
GAATCTGCATAGTCCA-1                 0.016515476
GCTTAAGCAAAGGTCG-1                 0.020222801
GTCACCTGTCAGGCTC-1                 0.025835081</code></pre>
</div>
</div>
<p>As there are only a very small number of cells classified as “platelets” (&lt; 20), it is difficult to figure out their precise cellular identity. Larger datasets would be required to confidently identify specific peaks for this population of cells, and further analysis performed to correctly annotate them. For downstream analysis we will thus remove the extremely rare cell states that were predicted, retaining only cell annotations with &gt;20 cells total.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>predicted_id_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(pbmc<span class="sc">$</span>predicted.id)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify the predicted.id values that have more than 20 cells</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>major_predicted_ids <span class="ot">&lt;-</span> <span class="fu">names</span>(predicted_id_counts[predicted_id_counts <span class="sc">&gt;</span> <span class="dv">20</span>])</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> pbmc[, pbmc<span class="sc">$</span>predicted.id <span class="sc">%in%</span> major_predicted_ids]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>For downstream analyses, we can simply reassign the identities of each cell from their UMAP cluster index to the per-cell predicted labels. It is also possible to consider merging the cluster indexes and predicted labels.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># change cell identities to the per-cell predicted labels</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(pbmc) <span class="ot">&lt;-</span> pbmc<span class="sc">$</span>predicted.id</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="step-9-rename-labels" class="level3" data-number="4.3.5">
<h3 data-number="4.3.5" class="anchored" data-anchor-id="step-9-rename-labels"><span class="header-section-number">4.3.5</span> Step 9: Rename labels</h3>
<p>Replace each cluster label with its most likely predicted label</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">levels</span>(pbmc)) {</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  cells_to_reid <span class="ot">&lt;-</span> <span class="fu">WhichCells</span>(pbmc, <span class="at">idents =</span> i)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  newid <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">which.max</span>(<span class="fu">table</span>(pbmc<span class="sc">$</span>predicted.id[cells_to_reid])))</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Idents</span>(pbmc, <span class="at">cells =</span> cells_to_reid) <span class="ot">&lt;-</span> newid</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="step-10-compare-the-results" class="level3" data-number="4.3.6">
<h3 data-number="4.3.6" class="anchored" data-anchor-id="step-10-compare-the-results"><span class="header-section-number">4.3.6</span> Step 10: Compare the results</h3>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># scRNA-seq</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(pbmc_rna, <span class="at">group.by =</span> <span class="st">"celltype"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"scRNA-seq"</span>)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene matrix</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>plot2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(pbmc, <span class="at">group.by =</span> <span class="st">"predicted.id"</span>, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"scATAC-seq (prediction)"</span>)</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Integration</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>plot3 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(pbmc, <span class="at">label =</span> T, <span class="at">group.by =</span> <span class="st">"ident"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"scATAC-seq (integration)"</span>)</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>plot1 <span class="sc">+</span> plot2 <span class="sc">+</span> plot3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="Section2_P14_intro_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>            used   (Mb) gc trigger    (Mb)   max used    (Mb)
Ncells  12536033  669.5   33923381  1811.8   53005282  2830.8
Vcells 910228372 6944.5 1656854019 12640.8 1425576862 10876.3</code></pre>
</div>
</div>
</section>
</section>
<section id="find-differentially-accessible-peaks-between-cell-types" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="find-differentially-accessible-peaks-between-cell-types"><span class="header-section-number">4.4</span> Find differentially accessible peaks between cell types</h2>
<p>In transcriptomic studies, we analyze differentially transcribed genes, so it is logical to study in ATAC-seq the genomic regions that are differentially accessible to the Tn5 transposase. To investigate differential chromatin accessibility, logistic regressions are used, as recommended by <a href="https://www.biorxiv.org/content/10.1101/258566v2">Ntranos <em>et al.</em> 2018</a>, and the total number of reads is included as a latent variable to mitigate the negative impact on results when dealing with libraries/samples with different sequencing depths.</p>
<p>A&nbsp;simple approach is to perform a Wilcoxon rank sum test, and the&nbsp;<a href="https://github.com/immunogenomics/presto">presto</a>&nbsp;package has implemented an extremely fast Wilcoxon test that can be run on a Seurat object.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"remotes"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">'remotes'</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">'immunogenomics/presto'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>For sparse data like scATAC-seq, it is necessary to adjust the <code>min.pct</code> parameter of the <code>FindMarkers()</code> function to lower values, as the default value (0.1) is designed for scRNA-seq data. Here we will focus on comparing Naive CD4 cells and CD14 monocytes, but any groups of cells can be compared using these methods. We can also visualize these marker peaks on a violin plot, feature plot, dot plot, heat map, or any&nbsp;<a href="https://satijalab.org/seurat/articles/visualization_vignette.html">visualization tool in Seurat</a>.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(presto)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="co"># change back to working with peaks instead of gene activities</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(pbmc) <span class="ot">&lt;-</span> <span class="st">'peaks'</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="co"># wilcox is the default option for test.use</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>da_peaks <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc,</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ident.1 =</span> <span class="st">"CD4 Naive"</span>,</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">ident.2 =</span> <span class="st">"CD14+ Monocytes"</span>,</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">test.use =</span> <span class="st">'wilcox'</span>,</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">min.pct =</span> <span class="fl">0.1</span></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(da_peaks)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                          p_val avg_log2FC pct.1 pct.2 p_val_adj
chr6-13302533-13303459        0  -5.308082 0.025 0.771         0
chr19-54207815-54208728       0  -4.370882 0.050 0.794         0
chr17-78198651-78199583       0  -5.653591 0.023 0.760         0
chr12-119988511-119989430     0   4.163235 0.782 0.090         0
chr7-142808530-142809435      0   3.665895 0.759 0.088         0
chr17-82126458-82127377       0   5.017039 0.699 0.042         0</code></pre>
</div>
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(da_peaks, <span class="at">file =</span> <span class="st">"data/da_peaks.RData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We visualize the results of the differential accessibility test using a violin plot and over the UMAP projection.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">VlnPlot</span>(</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc,</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> <span class="fu">rownames</span>(da_peaks)[<span class="dv">1</span>],</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">pt.size =</span> <span class="fl">0.1</span>,</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"CD4 Naive"</span>,<span class="st">"CD14+ Monocytes"</span>)</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>plot2 <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc,</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> <span class="fu">rownames</span>(da_peaks)[<span class="dv">1</span>],</span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">pt.size =</span> <span class="fl">0.1</span></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>plot1 <span class="sc">|</span> plot2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="Section2_P14_intro_files/figure-html/plot-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Peak coordinates can be difficult to interpret alone. We can find the closest gene to each of these peaks using the&nbsp;<a href="https://stuartlab.org/signac/reference/closestfeature"><code>ClosestFeature()</code></a>&nbsp;function.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>open_cd4naive <span class="ot">&lt;-</span> <span class="fu">rownames</span>(da_peaks[da_peaks<span class="sc">$</span>avg_log2FC <span class="sc">&gt;</span> <span class="dv">3</span>, ])</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>open_cd14mono <span class="ot">&lt;-</span> <span class="fu">rownames</span>(da_peaks[da_peaks<span class="sc">$</span>avg_log2FC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">3</span>, ])</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>closest_genes_cd4naive <span class="ot">&lt;-</span> <span class="fu">ClosestFeature</span>(pbmc, <span class="at">regions =</span> open_cd4naive)</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>closest_genes_cd14mono <span class="ot">&lt;-</span> <span class="fu">ClosestFeature</span>(pbmc, <span class="at">regions =</span> open_cd14mono)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co"># results</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(closest_genes_cd4naive)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                          tx_id    gene_name         gene_id   gene_biotype
ENSE00002206071 ENST00000397558       BICDL1 ENSG00000135127 protein_coding
ENST00000632998 ENST00000632998        PRSS2 ENSG00000275896 protein_coding
ENST00000583593 ENST00000583593       CCDC57 ENSG00000176155 protein_coding
ENST00000393054 ENST00000393054     ATP6V0A4 ENSG00000105929 protein_coding
ENST00000545320 ENST00000545320          CD6 ENSG00000013725 protein_coding
ENST00000509332 ENST00000509332 RP11-18H21.1 ENSG00000245954        lincRNA
                type            closest_region              query_region
ENSE00002206071 exon chr12-119989869-119990297 chr12-119988511-119989430
ENST00000632998  utr  chr7-142774509-142774564  chr7-142808530-142809435
ENST00000583593  gap   chr17-82108955-82127691   chr17-82126458-82127377
ENST00000393054  cds  chr7-138752625-138752837  chr7-138752283-138753197
ENST00000545320  gap   chr11-60971915-60987907   chr11-60985909-60986801
ENST00000509332  gap  chr4-152100818-152101483  chr4-152100248-152101142
                distance
ENSE00002206071      438
ENST00000632998    33965
ENST00000583593        0
ENST00000393054        0
ENST00000545320        0
ENST00000509332        0</code></pre>
</div>
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(closest_genes_cd14mono)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                          tx_id     gene_name         gene_id   gene_biotype
ENST00000606214 ENST00000606214        TBC1D7 ENSG00000145979 protein_coding
ENST00000448962 ENST00000448962          RPS9 ENSG00000170889 protein_coding
ENST00000592988 ENST00000592988         AFMID ENSG00000183077 protein_coding
ENSE00001638912 ENST00000455005 RP5-1120P11.3 ENSG00000231881        lincRNA
ENST00000635379 ENST00000635379     LINC01588 ENSG00000214900        lincRNA
ENSE00001389095 ENST00000340607         PTGES ENSG00000148344 protein_coding
                type           closest_region             query_region distance
ENST00000606214  gap   chr6-13267836-13305061   chr6-13302533-13303459        0
ENST00000448962  gap  chr19-54201610-54231740  chr19-54207815-54208728        0
ENST00000592988  gap  chr17-78191061-78202498  chr17-78198651-78199583        0
ENSE00001638912 exon   chr6-44073913-44074798   chr6-44058439-44059230    14682
ENST00000635379  gap  chr14-50039258-50039391  chr14-50038381-50039286        0
ENSE00001389095 exon chr9-129752887-129753047 chr9-129776928-129777838    23880</code></pre>
</div>
</div>
</section>
<section id="go-enrichment-analysis-with-clusterprofiler" class="level2" data-number="4.5">
<h2 data-number="4.5" class="anchored" data-anchor-id="go-enrichment-analysis-with-clusterprofiler"><span class="header-section-number">4.5</span> GO enrichment analysis with clusterProfiler</h2>
<p>We could follow up with this result by doing gene ontology enrichment analysis on the gene sets returned by&nbsp;<a href="https://stuartlab.org/signac/reference/closestfeature"><code>ClosestFeature()</code></a>,and there are many R packages that can do this (see the<a href="https://bioconductor.org/packages/release/bioc/html/GOstats.html"><code>GOstats</code></a>&nbsp;or&nbsp;<a href="https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html"><code>clusterProfiler</code></a>&nbsp;packages for example).</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(enrichplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>cd4naive_ego <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(<span class="at">gene =</span> closest_genes_cd4naive<span class="sc">$</span>gene_id, <span class="co"># like DEG</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">keyType =</span> <span class="st">"ENSEMBL"</span>, </span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">OrgDb =</span> org.Hs.eg.db, <span class="co"># organism</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">ont =</span> <span class="st">"BP"</span>, <span class="co"># Biological process</span></span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>, <span class="co"># Benjamini-Hochberg (BH)</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>,</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">qvalueCutoff =</span> <span class="fl">0.05</span>,</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">readable =</span> <span class="cn">TRUE</span>) <span class="co"># Convert the gene identifiers (ENSEMBL) to readable gene names.</span></span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(cd4naive_ego,<span class="at">showCategory =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="Section2_P14_intro_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>cd14mono_ego <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(<span class="at">gene =</span> closest_genes_cd14mono<span class="sc">$</span>gene_id,</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">keyType =</span> <span class="st">"ENSEMBL"</span>,</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">OrgDb =</span> org.Hs.eg.db,</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">ont =</span> <span class="st">"BP"</span>,</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>,</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>,</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">qvalueCutoff =</span> <span class="fl">0.05</span>,</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">readable =</span> <span class="cn">TRUE</span>)</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(cd14mono_ego,<span class="at">showCategory =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="Section2_P14_intro_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="plotting-genomic-regions" class="level2" data-number="4.6">
<h2 data-number="4.6" class="anchored" data-anchor-id="plotting-genomic-regions"><span class="header-section-number">4.6</span> Plotting genomic regions</h2>
<p>We can visualize the frequency of Tn5 integration across genomic regions for cells grouped by cluster, cell type, or any other metadata stored in the object using the <a href="https://stuartlab.org/signac/reference/coverageplot"><code>CoveragePlot()</code></a> function. These plots represent pseudo-bulk accessibility tracks, where the <strong>signal from all cells within a group is averaged to display DNA accessibility in a specific region.</strong> (Credit to Andrew Hill for the inspiration behind this function, as highlighted in his excellent <a href="http://andrewjohnhill.com/blog/2019/04/12/streamlining-scatac-seq-visualization-and-analysis/">blog post.</a>) In addition to accessibility tracks, we can include other key information such as gene annotations, peak coordinates, and genomic links (if available in the object). For further details, refer to the <a href="https://stuartlab.org/signac/articles/visualization">visualization vignette</a>.</p>
<p>For plotting purposes, it’s nice to have related cell types grouped together. We can automatically sort the plotting order according to similarities across the annotated cell types by running the&nbsp;<a href="https://stuartlab.org/signac/reference/sortidents"><code>SortIdents()</code></a>&nbsp;function:</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">SortIdents</span>(pbmc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Creating pseudobulk profiles for 13 variables across 165376 features</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing euclidean distance between pseudobulk profiles</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Clustering distance matrix</code></pre>
</div>
</div>
<p>We can then visualize the DA peaks open in CD4 naive cells and CD14 monocytes, near some key marker genes for these cell types, CD4 and LYZ respectively. Here we’ll highlight the DA peaks regions in grey.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># find DA peaks overlapping gene of interest</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>regions_highlight <span class="ot">&lt;-</span> <span class="fu">subsetByOverlaps</span>(<span class="fu">StringToGRanges</span>(open_cd4naive), <span class="fu">LookupGeneCoords</span>(pbmc, <span class="st">"CD4"</span>))</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a><span class="fu">CoveragePlot</span>(</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> pbmc,</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">region =</span> <span class="st">"CD4"</span>,</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">region.highlight =</span> regions_highlight,</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">extend.upstream =</span> <span class="dv">1000</span>,</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">extend.downstream =</span> <span class="dv">1000</span></span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 22 rows containing missing values or values outside the scale range
(`geom_segment()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 3 rows containing missing values or values outside the scale range
(`geom_segment()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 1 row containing missing values or values outside the scale range
(`geom_segment()`).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Section2_P14_intro_files/figure-html/plot coverage-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="calling-peaks" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Calling peaks</h1>
<p>You can call peaks on a single-cell ATAC-seq dataset using <a href="https://github.com/macs3-project/MACS"><strong>MACS2</strong></a>. To use this functionality in <strong>Signac</strong>, make sure MACS2 is installed—either through <a href="https://pypi.org/project/MACS2/">pip</a> or <a href="https://anaconda.org/bioconda/macs2">conda</a>, or by building it from <a href="https://github.com/macs3-project/MACS">source</a>.</p>
<p>For example, with scATAC-seq data from human PBMCs (as shown in our tutorial or from <a href="https://stuartlab.org/signac/articles/pbmc_vignette">Signac vignette</a>), you can load the necessary packages and a pre-computed Seurat object. See the vignette for the code used to generate this object and links to the raw data.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><a href="https://stuartlab.org/signac/articles/peak_calling"><img src="https://stuartlab.org/signac/articles/peak_calling_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img"></a></p>
</figure>
</div>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.4.1 (2024-06-14 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 22631)

Matrix products: default


locale:
[1] LC_COLLATE=English_United States.utf8 
[2] LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

time zone: America/Mexico_City
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] enrichplot_1.24.4         org.Hs.eg.db_3.19.1      
 [3] clusterProfiler_4.12.6    presto_1.0.0             
 [5] data.table_1.16.0         Rcpp_1.0.13              
 [7] future_1.34.0             EnsDb.Hsapiens.v86_2.99.0
 [9] ensembldb_2.28.1          AnnotationFilter_1.28.0  
[11] GenomicFeatures_1.56.0    AnnotationDbi_1.66.0     
[13] Biobase_2.64.0            patchwork_1.2.0          
[15] ggplot2_3.5.1             GenomicRanges_1.56.1     
[17] GenomeInfoDb_1.40.1       IRanges_2.38.1           
[19] S4Vectors_0.42.1          BiocGenerics_0.50.0      
[21] Seurat_5.1.0              SeuratObject_5.0.2       
[23] sp_2.1-4                  Signac_1.14.0            

loaded via a namespace (and not attached):
  [1] fs_1.6.4                    ProtGenerics_1.36.0        
  [3] matrixStats_1.4.0           spatstat.sparse_3.1-0      
  [5] bitops_1.0-8                httr_1.4.7                 
  [7] RColorBrewer_1.1-3          tools_4.4.1                
  [9] sctransform_0.4.1           utf8_1.2.4                 
 [11] R6_2.5.1                    lazyeval_0.2.2             
 [13] uwot_0.2.2                  withr_3.0.1                
 [15] gridExtra_2.3               progressr_0.14.0           
 [17] cli_3.6.2                   spatstat.explore_3.3-2     
 [19] fastDummies_1.7.4           scatterpie_0.2.4           
 [21] labeling_0.4.3              spatstat.data_3.1-2        
 [23] ggridges_0.5.6              pbapply_1.7-2              
 [25] Rsamtools_2.20.0            yulab.utils_0.1.7          
 [27] gson_0.1.0                  DOSE_3.30.5                
 [29] R.utils_2.12.3              parallelly_1.38.0          
 [31] limma_3.60.6                rstudioapi_0.16.0          
 [33] RSQLite_2.3.7               gridGraphics_0.5-1         
 [35] generics_0.1.3              BiocIO_1.14.0              
 [37] ica_1.0-3                   spatstat.random_3.3-1      
 [39] dplyr_1.1.4                 GO.db_3.19.1               
 [41] Matrix_1.7-0                ggbeeswarm_0.7.2           
 [43] fansi_1.0.6                 abind_1.4-8                
 [45] R.methodsS3_1.8.2           lifecycle_1.0.4            
 [47] yaml_2.3.10                 SummarizedExperiment_1.34.0
 [49] qvalue_2.36.0               SparseArray_1.4.8          
 [51] Rtsne_0.17                  grid_4.4.1                 
 [53] blob_1.2.4                  promises_1.3.0             
 [55] crayon_1.5.3                miniUI_0.1.1.1             
 [57] lattice_0.22-6              cowplot_1.1.3              
 [59] KEGGREST_1.44.1             pillar_1.9.0               
 [61] knitr_1.48                  fgsea_1.30.0               
 [63] rjson_0.2.22                future.apply_1.11.2        
 [65] codetools_0.2-20            fastmatch_1.1-4            
 [67] leiden_0.4.3.1              glue_1.7.0                 
 [69] ggfun_0.1.6                 spatstat.univar_3.0-1      
 [71] treeio_1.28.0               vctrs_0.6.5                
 [73] png_0.1-8                   spam_2.10-0                
 [75] gtable_0.3.5                cachem_1.1.0               
 [77] xfun_0.45                   S4Arrays_1.4.1             
 [79] mime_0.12                   tidygraph_1.3.1            
 [81] survival_3.6-4              RcppRoll_0.3.1             
 [83] statmod_1.5.0               fitdistrplus_1.2-1         
 [85] ROCR_1.0-11                 nlme_3.1-164               
 [87] ggtree_3.12.0               bit64_4.0.5                
 [89] RcppAnnoy_0.0.22            irlba_2.3.5.1              
 [91] vipor_0.4.7                 KernSmooth_2.23-24         
 [93] colorspace_2.1-1            DBI_1.2.3                  
 [95] ggrastr_1.0.2               tidyselect_1.2.1           
 [97] bit_4.0.5                   compiler_4.4.1             
 [99] curl_5.2.2                  httr2_1.0.5                
[101] DelayedArray_0.30.1         plotly_4.10.4              
[103] shadowtext_0.1.4            rtracklayer_1.64.0         
[105] scales_1.3.0                lmtest_0.9-40              
[107] rappdirs_0.3.3              stringr_1.5.1              
[109] digest_0.6.36               goftest_1.2-3              
[111] spatstat.utils_3.1-0        rmarkdown_2.28             
[113] XVector_0.44.0              htmltools_0.5.8.1          
[115] pkgconfig_2.0.3             MatrixGenerics_1.16.0      
[117] fastmap_1.2.0               rlang_1.1.3                
[119] htmlwidgets_1.6.4           UCSC.utils_1.0.0           
[121] shiny_1.9.1                 farver_2.1.2               
[123] zoo_1.8-12                  jsonlite_1.8.8             
[125] BiocParallel_1.38.0         GOSemSim_2.30.2            
[127] R.oo_1.26.0                 RCurl_1.98-1.16            
[129] magrittr_2.0.3              ggplotify_0.1.2            
[131] GenomeInfoDbData_1.2.12     dotCall64_1.1-1            
[133] munsell_0.5.1               ape_5.8                    
[135] viridis_0.6.5               reticulate_1.39.0          
[137] stringi_1.8.4               ggraph_2.2.1               
[139] zlibbioc_1.50.0             MASS_7.3-60.2              
[141] plyr_1.8.9                  parallel_4.4.1             
[143] listenv_0.9.1               ggrepel_0.9.5              
[145] deldir_2.0-4                Biostrings_2.72.1          
[147] graphlayouts_1.2.0          splines_4.4.1              
[149] tensor_1.5                  igraph_2.0.3               
[151] spatstat.geom_3.3-2         RcppHNSW_0.6.0             
[153] reshape2_1.4.4              XML_3.99-0.17              
[155] evaluate_1.0.0              tweenr_2.0.3               
[157] httpuv_1.6.15               RANN_2.6.2                 
[159] tidyr_1.3.1                 purrr_1.0.2                
[161] polyclip_1.10-7             scattermore_1.2            
[163] ggforce_0.4.2               xtable_1.8-4               
[165] restfulr_0.0.15             tidytree_0.4.6             
[167] RSpectra_0.16-2             later_1.3.2                
[169] viridisLite_0.4.2           tibble_3.2.1               
[171] aplot_0.2.3                 memoise_2.0.1              
[173] beeswarm_0.4.0              GenomicAlignments_1.40.0   
[175] cluster_2.1.6               globals_0.16.3             </code></pre>
</div>
</div>
<section id="references" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="references"><span class="header-section-number">5.1</span> References</h2>
<ul>
<li>Stuart, et al.&nbsp;2019. <a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867419305598%3Fshowall%3Dtrue">Comprehensive Integration of Single-Cell Data</a>. Cell.</li>
<li><a href="https://www.sc-best-practices.org/chromatin_accessibility/introduction.html">Single-cell ATAC sequencing</a></li>
<li>Heumos, et al . 2023. <a href="https://www.nature.com/articles/s41576-023-00586-w">Best practices for single-cell analysis across modalities</a>. Nature reviews genetics.</li>
<li>Signac tutorial - <a href="https://stuartlab.org/signac/articles/pbmc_vignette#create-a-gene-activity-matrix">Create a gene activity matrix</a></li>
<li>Signac tutorial - <a href="https://stuartlab.org/signac/articles/integrate_atac">scATAC-seq data integration</a></li>
<li>Signac tutorial - <a href="https://satijalab.org/seurat/articles/seurat5_atacseq_integration_vignette">Integrating scRNA-seq and scATAC-seq data</a></li>
<li>Signac tutorial - <a href="https://stuartlab.org/signac/articles/peak_calling">Calling peaks</a></li>
<li><a href="https://rpubs.com/gloknar/809238">Análisis de datos scATAC-seq en Signac: cerebro murino</a></li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./Section1_P13_retos.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Exercises</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./Section2_P14_retos.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Exercises</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>